features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularStack.cpp

```diff
@@ -0,0 +1,239 @@
+    _at.write_string(dataStr);
+    _at.cmd_stop();
+
+    _at.resp_start();
+    _at.skip_param(); // skip socket id
+    sent_len = _at.read_int();
+    _at.resp_stop();
+
+    free(dataStr);
+    if ((_at.get_last_error() == NSAPI_ERROR_OK)) {
+        return sent_len;
+    }
+
+    return _at.get_last_error();
+}
+
+nsapi_size_or_error_t UBLOX_N2XX_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+        void *buffer, nsapi_size_t size)
+{
+    nsapi_size_or_error_t nsapi_error_size = NSAPI_ERROR_DEVICE_ERROR;
+    nsapi_size_t read_blk, usorf_sz, count = 0, length = size*2;
+    bool success = true;
+    char ipAddress[NSAPI_IP_SIZE];
+    int port = 0;
+    Timer timer;
+
+    if (socket->pending_bytes == 0) {
+        _at.process_oob();
+        if (socket->pending_bytes == 0) {
+            return NSAPI_ERROR_WOULD_BLOCK;
+        }
+    }
+
+    timer.start();
+    while (success && (length > 0)) {
+        read_blk = N2XX_MAX_PACKET_SIZE;
+        if (read_blk > length) {
+            read_blk = length;
+        }
+        if (socket->pending_bytes > 0) {
+            _at.cmd_start("AT+NSORF=");
+            _at.write_int(socket->id);
+            _at.write_int(read_blk);
+            _at.cmd_stop();
+
+            _at.resp_start();
+            _at.skip_param(); // receiving socket id
+            _at.read_string(ipAddress, sizeof(ipAddress));
+            port = _at.read_int();
+            usorf_sz = _at.read_int();
+            if (usorf_sz > length) {
+                usorf_sz = length;
+            }
+            _at.read_hex_string((char *)buffer + count, usorf_sz*2+1);
+            _at.resp_stop();
+
+            // Must use what +NSORF returns here as it may be less or more than we asked for
+            if (usorf_sz >= socket->pending_bytes) {
+                socket->pending_bytes = 0;
+            } else {
+                socket->pending_bytes -= usorf_sz;
+            }
+
+            if (usorf_sz > 0) {
+                count += (usorf_sz*2);
+                length -= (usorf_sz*2);
+            } else {
+                // read() should not fail
+                success = false;
+            }
+        }  else if (timer.read_ms() < SOCKET_TIMEOUT) {
+            // Wait for URCs
+            _at.process_oob();
+        } else {
+            if (count == 0) {
+                // Timeout with nothing received
+                nsapi_error_size = NSAPI_ERROR_WOULD_BLOCK;
+                success = false;
+            }
+            length = 0; // This simply to cause an exit
+        }
+    }
+    timer.stop();
+
+    socket->pending_bytes = 0;
+    if (!count || (_at.get_last_error() != NSAPI_ERROR_OK)) {
+        return NSAPI_ERROR_WOULD_BLOCK;
+    }
+
+    if (success && socket->proto == NSAPI_UDP && address) {
+        address->set_ip_address(ipAddress);
+        address->get_ip_address();
+        address->set_port(port);
+    }
+
+    return nsapi_error_size = count/2;
```

Comment:
needs fix

Reply:
> @mudassar-ublox please check the remaining comments and it looks good :)
> 
> I noticed that `CellularSocket` is initialized incorrectly, and because of that `UBLOX_N2XX_CellularStack` probably fails on multiple sockets simultaneously. As a workaround please apply the patch below.
> 
> ```
> diff --git i/features/cellular/framework/AT/AT_CellularStack.cpp w/features/cellular/framework/AT/AT_CellularStack.cpp
> index 8c09f73849..f83259b16c 100644
> --- i/features/cellular/framework/AT/AT_CellularStack.cpp
> +++ w/features/cellular/framework/AT/AT_CellularStack.cpp
> @@ -134,9 +134,7 @@ nsapi_error_t AT_CellularStack::socket_open(nsapi_socket_t *handle, nsapi_protoc
>      _socket[index] = new CellularSocket;
>      CellularSocket *psock;
>      psock = _socket[index];
> -    memset(psock, 0, sizeof(CellularSocket));
>      SocketAddress addr(0, get_dynamic_ip_port());
> -    psock->id = index;
>      psock->localAddress = addr;
>      psock->proto = proto;
>      *handle = psock;
> ```
I think `psock->id = index;` is required, to identify sockets by its ID ?